File: 163_missing_ranges.java
class Solution {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            res.add(getRange((long)lower - 1, (long)upper + 1));
            return res;
        }
        
        for (int i = 0; i <= nums.length; i++) {
            String range = "";
            if (i == 0) {
                range = getRange((long)lower - 1, (long)nums[i]);
            } else if (i == nums.length) {
                range = getRange((long)nums[i - 1], (long)upper + 1);
            } else {
                range = getRange((long)nums[i - 1], (long)nums[i]);
            }
            if (! range.isEmpty()) {
                res.add(range);
            }
        }
        return res;
    }
    
    private String getRange(long left, long right) {
        long diff = right - left;
        if (diff <= 1) {
            return "";
        } else if (diff == 2) {
            return String.valueOf(left + 1);
        } else {
            return String.valueOf(left + 1) + "->" + String.valueOf(right - 1);
        }
    }
}

File: python/163_missing_ranges.py
class Solution:
    def findMissingRanges(self, nums: List[int], lower: int, upper: int) -> List[str]:
        res = []
        if not nums:
            self.addRange(lower, upper, res)
            return res

        if lower < nums[0]:
            self.addRange(lower, nums[0] - 1, res)

        for i in range(1, len(nums)):
            if nums[i] == nums[i - 1] + 1 or nums[i] == nums[i - 1]:
                continue
            else:
                self.addRange(nums[i - 1] + 1, nums[i] - 1, res)

        if nums[-1] < upper:
            self.addRange(nums[-1] + 1, upper, res)

        return res

    def addRange(self, lower: int, upper: int, res: List[str]):
        if lower == upper:
            res.append(str(lower))
        else:
            res.append(str(lower) + "->" + str(upper))
